### PR TITLE
Add GET endpoint to view details in a single record

### PR DESCRIPTION
**Pull request by LBHBomotoso**
Adds an endpoint to view the details within a single Resident Information object. 
The object is retrieved by its `academyId` (which would refer to a composite key made from of a `claim_id` value and a `person_ref` value).

**Any suggestions or comments on this PR would be appreciated.**

- [x] Return NI Number as part of the record.
- [x] Have `academyId` refer to a composite key of `claim_id` and `person_ref`
- [x] Have the ID attribute present in the Resident Information Object.
- [x] Edit the URI for the getbyid appropriately.

### CODE REVIEW

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj [hunk 1]
```diff
@@ -1,26 +1,25 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+<Project Sdk="Microsoft.NET.Sdk.Web">
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
 
     <IsPackable>false</IsPackable>
     <IsTestProject>true</IsTestProject>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
```
Comment (elcorbs) [inline AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:7]: I think this is the code analysis that @Renny4Real added, so don't think it should be deleted

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj [hunk 2]
```diff
@@ -1,26 +1,25 @@
...
-    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.0.0" />
     <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="3.1.4" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.3" />
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.9.0" />
-    <PackageReference Include="NUnit3TestAdapter" Version="3.12.0" />
-    <PackageReference Include="Bogus" Version="25.0.4" />
-    <PackageReference Include="Moq" Version="4.10.1" />
-    <PackageReference Include="nunit" Version="3.11.0" />
-    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="3.1.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.4" />
```
Comment (elcorbs) [inline AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:14]: We aren't using SqlServer so should be able to delete this

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj [hunk 3]
```diff
@@ -1,26 +1,25 @@
...
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.6.1" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.16.1">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="Bogus" Version="29.0.2" />
```
Comment (elcorbs) [inline AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:20]: Also don't think we are using Bogus (I realise these were already there but good to tidy up anyway :))

File: AcademyResidentInformationApi/AcademyResidentInformationApi.csproj [hunk 4]
```diff
@@ -6,22 +6,18 @@
...
-        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.HealthChecks" Version="1.0.0" />
     <PackageReference Include="Microsoft.AspNetCore.Mvc.Versioning" Version="4.1.1" />
     <PackageReference Include="Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer" Version="4.1.1" />
-    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.0.0" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.4" />
```
Comment (elcorbs) [inline AcademyResidentInformationApi/AcademyResidentInformationApi.csproj:19]: Again - we shouldn't need this anymore

File: AcademyResidentInformationApi/V1/UseCase/Interfaces/IGetResidentByIdUseCase.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+using AcademyResidentInformationApi.V1.Boundary.Responses;
+
+namespace AcademyResidentInformationApi.V1.UseCase.Interfaces
+{
+    public interface IGetResidentByIdUseCase
+    {
+        public ResidentInformation Execute(int id);
```
Comment (elcorbs) [inline AcademyResidentInformationApi/V1/UseCase/Interfaces/IGetResidentByIdUseCase.cs]: You don't need the `public` accessor in interfaces as everything is public

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 1 above)
Comment (LBHBomotoso) [reply AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:7]: It doesn't compile with the settings

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 2 above)
Comment (LBHBomotoso) [reply AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:14]: This was in the original project.

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 3 above)
Comment (LBHBomotoso) [reply AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:20]: This is in the original project

File: AcademyResidentInformationApi/AcademyResidentInformationApi.csproj (hunk 4 above)
Comment (LBHBomotoso) [reply AcademyResidentInformationApi/AcademyResidentInformationApi.csproj:19]: Updated as it is in the original project

File: AcademyResidentInformationApi/V1/UseCase/Interfaces/IGetResidentByIdUseCase.cs (hunk 5 above)
Comment (LBHBomotoso) [reply AcademyResidentInformationApi/V1/UseCase/Interfaces/IGetResidentByIdUseCase.cs]: true, same for the interface declaration?

Comment (elcorbs) [reply AcademyResidentInformationApi/V1/UseCase/Interfaces/IGetResidentByIdUseCase.cs]: I think you still need it there

**Review by humulla — COMMENTED** [review]
I think the overall changes made are great, I would agree to renaming residents to be claimant to make it more clear

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 1 above)
Comment (elcorbs) [inline AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:7]: Does this still need to be added back in?

File: AcademyResidentInformationApi/V1/Infrastructure/AcademyContext.cs [hunk 6]
```diff
@@ -15,15 +15,35 @@ public AcademyContext(DbContextOptions options) : base(options)
...
+            modelBuilder.Entity<Person>()
+                .HasKey(person => new
+                {
+                    person.Id,
+                    person.PersonRef
+                });
+
+            // composite foreign key for Address targeting unique columns in the Person table
+            modelBuilder.Entity<Person>()
+                .HasOne(x => x.Address)
```
Comment (elcorbs) [inline AcademyResidentInformationApi/V1/Infrastructure/AcademyContext.cs:36]: Is there definitely only one address record for each person?

Comment (Renny4Real) [reply AcademyResidentInformationApi/V1/Infrastructure/AcademyContext.cs:36]: I think it's one-to-one.
From what I remember, the Address and Person tables both have `HouseID` and `ClaimID` columns and you can join the tables on these columns e.g. 
```
JOIN hbhousehold ON hbmember.claim_id = hbhousehold.claim_id
		AND hbmember.house_id = hbhousehold.house_id
```
But I will go and confirm this. 👍

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 1 above)
Comment (Renny4Real) [reply AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:7]: Thanks for spotting this. I missed this while I was rebasing. 
Will add this back in.

**Comment by Renny4Real** [comment]
On hindsight, would renaming the Resident to Claimant clash with the work currently being done with the search parameters.
Maybe the rename should be done once the search parameters' changes are finished?

**Comment by humulla** [comment]
That's a good call - I think the renaming should be done once the search parameters task is complete

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 1 above)
Comment (Renny4Real) [reply AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:7]: Have added this in 86dd418

File: AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj (hunk 3 above)
Comment (louisebolden) [reply AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:20]: Do you mean the base-api project? Some things are still included there as options, which should be removed from templated/descendent projects if they do not need to use it.

Perhaps the `SqlServer` and `Bogus` packages here are examples of that, and should therefore indeed be removed here? (Certainly it seems like `SqlServer` should be.)

**Review by LBHBomotoso — COMMENTED** [review]
https://github.com/LBHackney-IT/academy-resident-information-api/pull/4/commits/7fb380f42d6e2c553fd2b9dabcc60c9b6b3c2ce1#diff-2a845cfdb562e239d9c04fa9df262507

Line 35 in this commit. We had to suppress the static analysis for string culture/comparison

File: database/schema.sql [hunk 7]
```diff
@@ -97,7 +58,48 @@ CREATE TABLE hbmember (
 	severely_mentally_impaired_ind smallint NULL,
 	period_sanction int NULL,
 	sanction_type int NULL,
-	sanction_rate_type int NULL
+	sanction_rate_type int NULL,
+	CONSTRAINT PK_hbmember PRIMARY KEY (claim_id, person_ref),
```
Comment (makeyw) [inline database/schema.sql:62]: Might be getting confused but person_ref isn't a PK in the Ingres table.
PK is made up of claim_id, house_id and member_id

Comment (Renny4Real) [reply database/schema.sql:62]: Yeah, I was told `person_ref` was part of the PK before I finding out that it wasn't in Ingres from our conversation.

Comment (makeyw) [reply database/schema.sql:62]: in which case you're perfectly correct....let's have that meeting :)

Comment (bjpirt) [reply database/schema.sql:62]: My bad! I thought it was but it looks like Single View is wrong

Comment (Renny4Real) [reply database/schema.sql:62]: No worries!

**Comment by Renny4Real** [comment]
Merging this as per team discussion so that we can progress with fleshing out the API functionality and renaming the endpoint from `/resident` to `/claimant`. See #17.
Any changes needed will be addressed in a separate PR. 😄

**Pull request merged**